mmdeploy/codebase/mmdet3d/deploy/mono_detection.py

```diff
@@ -0,0 +1,207 @@
+# Copyright (c) OpenMMLab. All rights reserved.
+import os
+from copy import deepcopy
+from typing import Dict, Optional, Sequence, Tuple, Union
+
+import mmengine
+import numpy as np
+import torch
+from mmdet3d.structures import get_box_type
+from mmengine import Config
+from mmengine.dataset import Compose, pseudo_collate
+from mmengine.model import BaseDataPreprocessor
+
+from mmdeploy.codebase.base import BaseTask
+from mmdeploy.utils import Task
+from .mmdet3d import MMDET3D_TASK
+
+
+def _get_dataset_metainfo(model_cfg: Config):
+    """Get metainfo of dataset.
+
+    Args:
+        model_cfg Config: Input model Config object.
+
+    Returns:
+        list[str]: A list of string specifying names of different class.
+    """
+
+    for dataloader_name in [
+            'test_dataloader', 'val_dataloader', 'train_dataloader'
+    ]:
+        if dataloader_name not in model_cfg:
+            continue
+        dataloader_cfg = model_cfg[dataloader_name]
+        dataset_cfg = dataloader_cfg.dataset
+        if 'metainfo' in dataset_cfg:
+            return dataset_cfg.metainfo
+    return None
+
+
+@MMDET3D_TASK.register_module(Task.MONO_DETECTION.value)
+class MonoDetection(BaseTask):
+
+    def __init__(self, model_cfg: mmengine.Config, deploy_cfg: mmengine.Config,
+                 device: str):
+        super().__init__(model_cfg, deploy_cfg, device)
+
+    def build_backend_model(self,
+                            model_files: Sequence[str] = None,
+                            **kwargs) -> torch.nn.Module:
+        """Initialize backend model.
+
+        Args:
+            model_files (Sequence[str]): Input model files.
+
+        Returns:
+            nn.Module: An initialized backend model.
+        """
+        from .mono_detection_model import build_mono_detection_model
+
+        data_preprocessor = deepcopy(
+            self.model_cfg.model.get('data_preprocessor', {}))
+        data_preprocessor.setdefault('type', 'mmdet3D.Det3DDataPreprocessor')
+
+        model = build_mono_detection_model(
+            model_files,
+            self.model_cfg,
+            self.deploy_cfg,
+            device=self.device,
+            data_preprocessor=data_preprocessor)
+        model = model.to(self.device)
+        return model
+
+    def create_input(
+        self,
+        pcd: Union[str, Sequence[str]],
```

Comment:
Could we feed input of `jpg/png` images instead of `.pkl` files?

Reply:
I think the information in pkl file is used for visualize 3d box in image, since 2d needs information to map back to 3d.

Those codes refer to  https://github.com/open-mmlab/mmdetection3d/blob/main/mmdet3d/apis/inference.py#L299 in function `def inference_mono_3d_detector`.

```
        mono_img_info = {f'{cam_type}': data_info['images'][cam_type]}
        data_ = dict(
            images=mono_img_info,
            box_type_3d=box_type_3d,
            box_mode_3d=box_mode_3d)

        data_ = test_pipeline(data_)
```
The `test_pipeline` needs mono_img_info from pkl file.

Also refer to doc https://github.com/open-mmlab/mmdetection3d/blob/main/docs/zh_cn/advanced_guides/datasets/nuscenes.md `info['images']：是一个字典，包含与每个相机对应的六个键值：'CAM_FRONT', 'CAM_FRONT_RIGHT', 'CAM_FRONT_LEFT', 'CAM_BACK', 'CAM_BACK_LEFT', 'CAM_BACK_RIGHT'。每个字典包含了对应相机的所有数据信息。`

Reply:
@zhjunqin hi, could we test whether files from kitti dataset could be used here? since the model is trained on kitti dataset: https://github.com/open-mmlab/mmdetection3d/blob/main/configs/smoke/smoke_dla34_dlaneck_gn-all_4xb8-6x_kitti-mono3d.py

Reply:
Ok, let me see.

Reply:
I checked dataset https://github.com/open-mmlab/mmdetection3d/blob/main/configs/_base_/datasets/kitti-mono3d.py which used by [smoke_dla34_dlaneck_gn-all_4xb8-6x_kitti-mono3d.py](https://github.com/open-mmlab/mmdetection3d/blob/main/configs/smoke/smoke_dla34_dlaneck_gn-all_4xb8-6x_kitti-mono3d.py).
I found that the dataset [LoadImageFromFileMono3D](https://github.com/open-mmlab/mmdetection3d/blob/main/mmdet3d/datasets/transforms/loading.py#L222) must needs a pkl file and an image.

If use KITTI dataset, then need a KITTI pkl file and a png file, like files listed in https://github.com/open-mmlab/mmdetection3d/tree/main/demo/data/kitti.
The [000008.pkl](https://github.com/open-mmlab/mmdetection3d/blob/main/demo/data/kitti/000008.pkl) includes informat about 3d, the [000008.png](https://github.com/open-mmlab/mmdetection3d/blob/main/demo/data/kitti/000008.png) is the 2d image.
Checked https://github.com/open-mmlab/mmdetection3d/blob/main/docs/zh_cn/advanced_guides/datasets/kitti.md for detail `kitti_infos_train.pkl：训练数据集，该字典包含了两个键值：metainfo 和 data_list。metainfo 包含数据集的基本信息`.

Also checked the code LoadImageFromFileMono3D, we can see from comments.
```
@TRANSFORMS.register_module()
class LoadImageFromFileMono3D(LoadImageFromFile):
    """Load an image from file in monocular 3D object detection. Compared to 2D
    detection, additional camera parameters need to be loaded.

    Args:
        kwargs (dict): Arguments are the same as those in
            :class:`LoadImageFromFile`.
    """

    def transform(self, results: dict) -> dict:
        """Call functions to load image and get image meta information.

        Args:
            results (dict): Result dict from :obj:`mmdet.CustomDataset`.

        Returns:
            dict: The dict contains loaded image and meta information.
        """
        # TODO: load different camera image from data info,
        # for kitti dataset, we load 'CAM2' image.
        # for nuscenes dataset, we load 'CAM_FRONT' image.

        if 'CAM2' in results['images']:
            filename = results['images']['CAM2']['img_path']
            results['cam2img'] = results['images']['CAM2']['cam2img']
        elif len(list(results['images'].keys())) == 1:
            camera_type = list(results['images'].keys())[0]
            filename = results['images'][camera_type]['img_path']
            results['cam2img'] = results['images'][camera_type]['cam2img']
```

Reply:
I checked mmdeploy, seems there is no kitti pkl and image file?

Reply:
I tested with kitti dataset, it works with following codes.

```
        for cam_type, item in images.items():
            # avoid data_info['images'] has multiple keys and camera views.
            # Only support one cam type here
            if not cam_type in ['CAM_FRONT', 'CAM2']:
                continue
```